torch/distributions/transforms.py

```diff
@@ -0,0 +1,235 @@
+    # Look up by Constraint subclass.
+    try:
+        Trans = _TRANSFORMS[type(constraint)]
+    except KeyError:
+        raise NotImplementedError(
+            'Cannot transform {} constraints'.format(type(constraint).__name__))
+    return Trans(constraint)
+
+
+def register_transform(constraint):
+    """
+    Decorator to register a :class:`~torch.distributions.constraints.Constraint`
+    subclass or singleton object with the :meth:`transform` function. Usage::
+
+        @register_transform(MyConstraintClass)
+        class MyTransform(Transform):
+            def to_unconstrained(self, x):
+                ...
+            def to_constrained(self, u):
+                ...
+
+    Args:
+        constraint (:class:`Constraint` subclass or instance):
+            Either a specific constraint or a class of constraints.
+    """
+
+    def decorator(transform_class):
+        if isinstance(constraint, constraints.Constraint):
+            # Register singleton instances.
+            _TRANSFORMS[constraint] = transform_class(constraint)
+        elif issubclass(constraint, constraints.Constraint):
+            # Register Constraint subclass.
+            _TRANSFORMS[constraint] = transform_class
+        else:
+            raise TypeError('Expected constraint to be either a Constraint subclass or instance, '
+                            'but got {}'.format(constraint))
+        return transform_class
+
+    return decorator
+
+
+class Transform(object):
+    """
+    Each constraint class registers a pseudoinverse pair of transforms
+    :meth:`~Transform.to_unconstrained` and :meth:`~Transform.to_constrained`.
+    These allow standard parameters to be transformed to an unconstrained space
+    for optimization and transformed back after optimization. Note that these
+    are not necessarily inverse pairs since the unconstrained space may have
+    extra dimensions that are projected out; only the one-sided inverse
+    equation is guaranteed::
+
+        x == t.to_constrained(t.to_unconstrained(x))
+
+    Args:
+        constraint (:class:`~torch.distributions.constraints.Constraint`):
+            A constraint object.
+    """
+    def __init__(self, constraint):
+        self.constraint = constraint
+
+    def to_unconstrained(self, x):
+        """
+        Transform from constrained coordinates to unconstrained coordinates.
+
+        Args:
+            x (Variable or Tensor): A constrained value.
+
+        Returns:
+            (Variable or Tensor): An unconstrained value.
+        """
+        raise NotImplementedError
+
+    def to_constrained(self, u):
+        """
+        Transform from unconstrained coordinates to constrained coordinates.
+
+        Args:
+            u (Variable or Tensor): An unconstrained value.
+
+        Returns:
+            (Variable or Tensor): A constrained value.
+        """
+        raise NotImplementedError
+
+
+@register_transform(constraints.real)
+class IdentityTransform(Transform):
```

Comment:
What's the difference between, e.g., `IdentityTransform` and an `IdentityBijector`? In general, are there left-inverse-only transforms that don't have a log det Jacobian we can implement as a Bijector?

Reply:
`Transform` is just a much lighter weight thing with fewer guarantees than `Bijector`. The main examples are manifolds that are more naturally represented as subspaces than as a single chart. E.g. von-Mises or spherical von-Mises variables can be represented as 2 and 3-vectors rather than unrolled into a local angle. Less contrived is the `simplex` constraint. We usually use `softmax` to optimize an `n+1`-dimensional parameter space and then project (not biject) down to an `n` dimensional constrained space.

Maybe I'm taking `bijector` too literally? Am I missing something?

Reply:
I think that design decision makes sense. I guess in many cases the transform could be refactored to simply call a Bijector's forward and inverse under the hood. But in other settings, it can't be written as a Bijector (if we define a Bijector as a class that requires implementing forward, inverse, and log det Jacobian).

Reply:
That's a great point: this could be a thin interface with only one or two non-`Bijector` pairs.